src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+            if checker.last_l0_bytes != 0 {
+                let l0_flow = checker.last_l0_bytes as f64
+                    / checker.last_l0_bytes_time.saturating_elapsed_secs();
+                checker.last_l0_bytes_time = Instant::now_coarse();
+                checker.short_term_l0_flow.observe(l0_flow as u64);
+                SCHED_L0_FLOW_GAUGE
+                    .with_label_values(&[&cf])
+                    .set(checker.short_term_l0_flow.get_avg() as i64);
+            }
+
+            checker.last_flush_bytes_time = Instant::now_coarse();
+            checker.last_l0_bytes = 0;
+            checker.last_flush_bytes = 0;
+
+            if checker.on_start_l0_files {
+                if num_l0_files < self.l0_files_threshold
+                    || checker.long_term_num_l0_files.trend() == Trend::Increasing
+                {
+                    // the write is accumulating, still need to throttle
+                    checker.on_start_l0_files = false;
+                } else {
+                    // still on start, should not throttle now
+                    return;
+                }
+            }
+
+            if let Some(throttle_cf) = self.throttle_cf.as_ref() {
+                if &cf != throttle_cf {
+                    return;
+                }
+            }
+
+            // adjust throttle speed based on flush flow and target flow
+            if let Some(_num_l0_for_last_update_target_flow) =
+                self.num_l0_for_last_update_target_flow
+            {
+                if self.cf_checkers[&cf].long_term_flush_flow.get_avg() > self.l0_target_flow
+                    && self.cf_checkers[&cf].short_term_flush_flow.get_recent() as f64
+                        > self.l0_target_flow
+                {
+                    SCHED_THROTTLE_ACTION_COUNTER
+                        .with_label_values(&[&cf, "down_flow"])
+                        .inc();
+                    self.down_flow(cf);
+                } else if (self.cf_checkers[&cf].short_term_flush_flow.get_avg()
+                    < self.l0_target_flow
+                    || (self.cf_checkers[&cf].short_term_flush_flow.get_recent() as f64)
+                        < self.l0_target_flow)
+                    && self.write_flow_recorder.get_recent() as f64
+                        > self.limiter.speed_limit() * 0.95
+                {
+                    SCHED_THROTTLE_ACTION_COUNTER
+                        .with_label_values(&[&cf, "up_flow"])
+                        .inc();
+                    self.up_flow(cf);
+                } else {
+                    SCHED_THROTTLE_ACTION_COUNTER
+                        .with_label_values(&[&cf, "keep_flow"])
+                        .inc();
+                }
+            } else {
+                SCHED_THROTTLE_ACTION_COUNTER
+                    .with_label_values(&[&cf, "no_target_flow"])
+                    .inc();
+            }
+        }
+    }
+
+    fn up_flow(&mut self, cf: String) {
```

Comment:
```suggestion
    fn increase_limit_for_l0(&mut self, cf: String) {
```

Reply:
And why L0 uses a much more sophisticated algorithm than memtable throttle?

Reply:
memtables can be consumed one by one, but all L0 files are involved in one compaction due to range overlap, so #L0 is of great unsteadiness which complicates the algorithm much.